test/csuite/wt8659_reconstruct_database_from_logs/main.c

```diff
@@ -0,0 +1,330 @@
+        testutil_check(system(buf));
+        printf("\t Table \"%s\": OK\n", uri);
+    }
+
+    return (ret);
+}
+
+/*
+ * add_work --
+ *     Inserts some data into the database.
+ */
+static void
+add_work(int iter)
+{
+    WT_CURSOR *cursor;
+    int i;
+    char k[32], v[128];
+
+    testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
+    /* Perform some operations with individual auto-commit transactions. */
+    for (i = 0; i < MAX_KEYS; i++) {
+        testutil_check(__wt_snprintf(k, sizeof(k), "key.%d.%d", iter, i));
+        testutil_check(__wt_snprintf(v, sizeof(v), "value.%d.%d", iter, i));
+        cursor->set_key(cursor, k);
+        cursor->set_value(cursor, v);
+        testutil_check(cursor->insert(cursor));
+    }
+    testutil_check(cursor->close(cursor));
+}
+
+/*
+ * take_full_backup --
+ *     Takes full backup of the database.
+ */
+static void
+take_full_backup(const char *home, const char *backup_home)
+{
+    WT_CURSOR *cursor;
+    int ret;
+    char buf[1024];
+    const char *filename;
+
+    testutil_check(session->open_cursor(session, "backup:", NULL, NULL, &cursor));
+
+    while ((ret = cursor->next(cursor)) == 0) {
+        testutil_check(cursor->get_key(cursor, &filename));
+        testutil_check(
+          __wt_snprintf(buf, sizeof(buf), "cp %s/%s %s/%s", home, filename, backup_home, filename));
+        testutil_check(system(buf));
+    }
+
+    scan_end_check(ret == WT_NOTFOUND);
+    testutil_check(cursor->close(cursor));
+}
+
+/*
+ * take_incr_backup --
+ *     Takes incremental log-based backup of the database.
+ */
+static void
+take_incr_backup(const char *backup_home, bool truncate_logs)
+{
+    WT_CURSOR *cursor;
+    int ret;
+    char buf[1024];
+    const char *filename;
+
+    testutil_check(session->open_cursor(session, "backup:", NULL, "target=(\"log:\")", &cursor));
+
+    while ((ret = cursor->next(cursor)) == 0) {
+        testutil_check(cursor->get_key(cursor, &filename));
+
+        testutil_check(__wt_snprintf(
+          buf, sizeof(buf), "cp %s/%s %s/%s", home_live, filename, backup_home, filename));
+        testutil_check(system(buf));
+    }
+    scan_end_check(ret == WT_NOTFOUND);
+
+    if (truncate_logs) {
+        /*
+         * With an incremental cursor, we want to truncate on the backup cursor to archive the logs.
+         * Only do this if the copy process was entirely successful.
+         */
+        testutil_check(session->truncate(session, "log:", cursor, NULL, NULL));
+    }
+
+    testutil_check(cursor->close(cursor));
+}
+
+/*
+ * prepare_folders --
+ *     Prepares all working folders required for the test.
+ */
+static void
+prepare_folders(void)
+{
+    reset_dir(home_live);
+    reset_dir(home_full);
+    reset_dir(home_incr);
+    reset_dir(home_incr_copy);
+}
+
+/*
+ * cleanup --
+ *     Test's cleanup.
+ */
+static void
+cleanup(void)
+{
+    testutil_check(conn->close(conn, NULL));
```

Comment:
Should we check that `conn` is not NULL before?

Reply:
If `conn` is NULL the program would not reach this method and crash earlier in `reopen_conn`.